docs/proposals/status.md

```diff
@@ -0,0 +1,300 @@
+
+### Conditions
+[Conditions](https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#typical-status-properties) are used to describe the status of resources in an extensible way. Out of tree controllers support a variety of conditions. There is no consistent usage across the core controller. This is exacerbated with custom controllers. To tackle this diversity and its impact in status computation for Application CRD, we need a short term and a long term solution. Short term solution is to do aggregation of diverse core resources in the Application CRD controller. Long term is to identify best practices and standardize conditions and propagate them to the core controllers thereby making Application CRD aggregation simpler.
+
+#### Standard Conditions
+These conditions should be added to `.status.conditions` for Application CRD and subsequently to other resources:
+##### `Ready`
+Readiness as perceived by the resource controller. This means the controller deems the underlying resources to be ready for consumption to the best of its knowledge. It coule be implemented as a low pass filter to filter out transient or persistent failures such as restarts of underlying pods.
+
+##### `Settled`
+Would indicate whether controller acted on latest revision of `.spec` and the underlying resources conform to the latest revision of the `.spec`
+
+##### `Error`
+Would capture Errors seen in last reconcilation loop.
+
+#### Other Conditions from core controllers
+These are not as universal for all kubernetes objects as `Ready`, `Settled`, `Error` and provide yet another dimension of `Ready` and `Settled` condition. 
+
+##### `Progressing`
+Deployment uses this to indicate there is an update to the underlying ReplicaSet that is being acted upon. This is a inversion of the `Settled` condition.
+
+##### `Available`
+Deployment uses this to indicate that underlying pods have been `Ready` for a preconfigured amount of time. This is not a strong heuristic for actual availability but rather a low pass filter that fails when pods are restarting continiously.
+
+#### Other Conditions
+##### `Disruption`
+`Disruption` is a special case of `Error` where we may know what caused the error when `Ready` is false. It is useful for some resources and could be standardized as an optional condition. 
+
+### Aggregating status
+#### Custom Resources
+For custom resources, look for `Ready`, `Error` and `Settled`. If the object does not have the necessary conditions mark that objects status as "Unknown" and would not be used for computing the aggregate status. This would cover the case for nested Applciation CRDs as well. Based on feedback from users and reviewers this stance could be updated.
+
+### Core controller resources
```

Comment:
For some resources, it is impossible to decide readiness from the resource definition itself. `Pod` is the best example of this, because (_e.g._) we don't know statically whether it is supposed to complete or not. It would be very nice to be able to display intent using annotations, or a similar mechanism. In the case of `Pod` itself, consider that Helm's test hooks allow us to launch `Pod`s that test whether an application is running -- so it is useful to (_e.g._) a monitoring/CI/CD system to communicate this intent, as it would allow it to wait for the tests to complete before progressing.

Reply:
Yeah, I personally would much rather use annotations instead of shared `.status` mutation. :)

Reply:
Ack.